Harden batchlog_manager stop and call from main in deferred action

bhalevy opened the pull request:
This PR contains the parts relevant to batchlog_manager stop in #8998 without adding a gate to the storage_proxy for synchronization with on-going queries in storage_proxy::drain_on_shutdown.

As explained in #9009, we see that the batchlog_manager isn't stopped if scylla shuts down during startup, e.g. when waiting for gossip to settle, since currently the batchlog_manager is stopped only from `storage_service::do_drain`, while `storage_service::drain_on_shutdown` deferred shutdown is installed only later on:
https://github.com/scylladb/scylla/blob/222ef17305b07e8faabd5d921e2bad0d7f78e106/main.cc#L1419-L1421

Fixes #9009

Test: unit(dev)
DTest: compact_storage_tests.py:TestCompactStorage.wide_row_test paging_test:TestPagingDatasetChanges.test_cell_TTL_expiry_during_paging update_cluster_layout_tests:TestUpdateClusterLayout.simple_add_new_node_while_adding_info_{1,2}_test (dev)

avikivity commented on db/batchlog_manager.cc:
```diff
@@ -123,7 +123,9 @@ future<> db::batchlog_manager::start() {
             (void)do_batch_log_replay().handle_exception([] (auto ep) {
```
I think there's still a problem here. do_batch_log_replay doesn't enter the gate, so it won't synchronize with stop().

bhalevy replied:
That's exactly the reason I added `_sem.wait()` to stop() in https://github.com/scylladb/scylla/pull/9010/commits/19e026ca3b5e94eaa755f7a064f97a5b0f84ee19

bhalevy replied:
@avikivity ping, please re-review.

avikivity commented:
I think a better way than having a timer start itself again recursively is for start() to launch a coroutine (storing its returned future somewhere) and for stop() to tell the coroutine to stop. Right now we have a mix of semaphores, gates, and checks of _stop, together with cross-shard calls. It's unlikely to work reliably.

But let's fix it first and then rewrite it.

bhalevy commented:
> I think a better way than having a timer start itself again recursively is for start() to launch a coroutine (storing its returned future somewhere) and for stop() to tell the coroutine to stop. Right now we have a mix of semaphores, gates, and checks of _stop, together with cross-shard calls. It's unlikely to work reliably.

Sounds simpler indeed.

> 
> But let's fix it first and then rewrite it.

xemul commented on main.cc:
```diff
@@ -1277,6 +1277,11 @@ int main(int ac, char** av) {
+            auto stop_batchlog_manager = defer_verbose_shutdown("batchlog manager", [] {
+                db::get_batchlog_manager().invoke_on_all([] (auto& bm) {
+                    return bm.stop();
+                }).get();
```
nit: Can be one-liner
```c++
db::get_batchlog_manager().invoke_on_all(&db::batchlog_manager::stop).get();
```

bhalevy replied:
fixed in v2.

xemul commented on db/batchlog_manager.cc:
```diff
@@ -140,7 +140,13 @@ future<> db::batchlog_manager::stop() {
+    // Synchronize with do_batch_log_replay.
+    // Although _sem might be held only on shard 0
+    // it is safe to wait for it on all shards.
+    return _sem.wait().then([this] {
```
Maybe teach `do_batchlog_replay()` run inside the `_gate` instead? Thus the `_gate.close()` will capture those too.

bhalevy replied:
solved in a better way in v2 by waiting on _started future returned from the replay loop after requesting abort.

bhalevy commented:
In v2:
- Harden start/stop by using an abort_source to abort from the replay loop.
- Extract the loop into batchlog_replay_loop() coroutine, with the _stop abourt source as a stop condition,
    plus use it for sleep_abortable to be able to promptly stop while sleeping.
- start() stores its returned future in an optional _started member, which is waited on in stop() to synchronize with the start process at any stage.
- stop() was made idempotent by making sure each required step (request_abort, wait on _strated, and closing _gate) is done only once.
- simplified batchlog_manager stop using `invoke_on_all(&db::batchlog_manager::stop)`

bhalevy commented:
In v3:
- updated the comment for batchlog_replay_loop (no _timer anymore)
- batchlog_manager: do_batch_log_replay: hold _gate
- batchlog_manager: make it peering_sharded_service
- batchlog_manager: stop: break _sem on shard 0

bhalevy commented:
ping review

avikivity commented on db/batchlog_manager.cc:
```diff
@@ -110,35 +112,47 @@ future<> db::batchlog_manager::do_batch_log_replay() {
 }
 
+// Since replay is a "node global" operation, we should not attempt to do
+// it in parallel on each shard. It will just overlap/interfere.  To
```
I don't agree with this moved comment.

bhalevy replied:
You don't agree with its contents or it being moved?
Changing this part is out of scope for this fix...

avikivity commented on db/batchlog_manager.hh:
```diff
@@ -89,12 +90,12 @@ private:
     db_clock::duration _write_request_timeout;
     uint64_t _replay_rate;
-    timer<clock_type> _timer;
+    std::optional<future<>> _started;
```
A simpler way is to use a non-optional future intialized to the coroutine, or to a ready future. But this is fine too.

bhalevy replied:
okay, will use a plain future<> initialized to a ready one.

avikivity commented on main.cc:
```diff
@@ -1277,6 +1277,9 @@ int main(int ac, char** av) {
             }).get();
+            auto stop_batchlog_manager = defer_verbose_shutdown("batchlog manager", [] {
+                db::get_batchlog_manager().invoke_on_all(&db::batchlog_manager::stop).get();
+            });
```
stop() is not idempotent due to the std::exchange(). The first stop() will wait, the second stop() will not.

If you want to make stop() idempotent, use a shared_future. The best is not to rely on stop() idempotency, it's just tricky. Call stop() exactly one time, and abort() to abort the loop.

bhalevy replied:
okay, makes sense

bhalevy replied:
fixed in v4

avikivity commented on db/batchlog_manager.cc:
```diff
@@ -94,22 +94,22 @@ db::batchlog_manager::batchlog_manager(cql3::query_processor& qp, batchlog_manag
 future<> db::batchlog_manager::do_batch_log_replay() {
     // Use with_semaphore is much simpler, but nested invoke_on can
     // cause deadlock.
-    return get_batchlog_manager().invoke_on(0, [] (auto& bm) {
```
Changelog lacks motivation.

bhalevy replied:
I'll send it separately as a followup.
It's immaterial to this fix...

bhalevy commented:
In v4:
- reordered the patches
- dropped "batchlog_manager: make it peering_sharded_service"
- refactored `drain` out of `stop` to abort the replay loop and wait on it.
  - it enters the batchlog manager `_gate` so it can be waited on by stop()
- existing calls to stop() changed to drain()
- stop() is called once using a deferred action in main.
